File: src-tauri/src/commands/files.rs
//! File system commands

use snowflake_core::{FileEntry, FileInfo, Result};

/// Read file contents as a string
#[tauri::command]
pub async fn read_file(path: String) -> Result<String> {
    snowflake_fs::read_file(&path).await
}

/// Write content to a file
#[tauri::command]
pub async fn write_file(path: String, content: String) -> Result<()> {
    snowflake_fs::write_file(&path, &content).await
}

/// List directory contents
#[tauri::command]
pub async fn list_directory(path: String) -> Result<Vec<FileEntry>> {
    snowflake_fs::list_directory(&path).await
}

/// Delete a file or directory
#[tauri::command]
pub async fn delete_file(path: String) -> Result<()> {
    snowflake_fs::delete_file(&path).await
}

/// Rename or move a file
#[tauri::command]
pub async fn rename_file(old_path: String, new_path: String) -> Result<()> {
    snowflake_fs::rename_file(&old_path, &new_path).await
}

/// Create a directory
#[tauri::command]
pub async fn create_directory(path: String) -> Result<()> {
    snowflake_fs::create_directory(&path).await
}

/// Check if a file exists
#[tauri::command]
pub async fn file_exists(path: String) -> bool {
    snowflake_fs::file_exists(&path).await
}

/// Get detailed file information
#[tauri::command]
pub async fn get_file_info(path: String) -> Result<FileInfo> {
    snowflake_fs::get_file_info(&path).await
}

File: apps/agent/src/lib/api/files.ts
/**
 * File Operations
 *
 * Functions for reading, writing, and managing files and directories.
 */

import { invoke, listen } from './core';

// ============================================
// Types
// ============================================

export interface FileEntry {
  /** Full path to the file */
  path: string;
  /** File name without directory */
  name: string;
  /** Whether this is a directory */
  isDir: boolean;
  /** Whether this is a symbolic link */
  isSymlink: boolean;
  /** Whether this is a hidden file (name starts with dot) */
  isHidden: boolean;
  /** Whether this file is ignored by git (.gitignore) */
  isGitIgnored: boolean;
  /** File size in bytes (undefined for directories) */
  size?: number;
  /** Last modified timestamp (Unix epoch seconds) */
  modified?: number;
}

export interface FileInfo {
  path: string;
  name: string;
  isDir: boolean;
  isFile: boolean;
  size: number;
  modified: number;
  created: number;
  readonly: boolean;
}

export interface FileChangeEvent {
  type: 'created' | 'modified' | 'deleted' | 'renamed';
  path: string;
  newPath?: string;
}

// ============================================
// File Operations
// ============================================

export async function readFile(path: string): Promise<string> {
  return invoke<string>('read_file', { path });
}

export async function readFileBytes(path: string): Promise<number[]> {
  return invoke<number[]>('read_file_bytes', { path });
}

export async function writeFile(path: string, content: string): Promise<void> {
  return invoke('write_file', { path, content });
}

export async function writeFileBytes(path: string, content: number[]): Promise<void> {
  return invoke('write_file_bytes', { path, content });
}

export async function listDirectory(path: string, showHidden?: boolean): Promise<FileEntry[]> {
  return invoke<FileEntry[]>('list_directory', { path, showHidden });
}

export async function createFile(path: string): Promise<void> {
  return invoke('create_file', { path });
}

export async function createDirectory(path: string): Promise<void> {
  return invoke('create_directory', { path });
}

export async function deleteFile(path: string): Promise<void> {
  return invoke('delete_file', { path });
}

export async function renameFile(oldPath: string, newPath: string): Promise<void> {
  return invoke('rename_file', { oldPath, newPath });
}

export async function copyFile(from: string, to: string): Promise<void> {
  return invoke('copy_file', { from, to });
}

export async function revealInFileManager(path: string): Promise<void> {
  return invoke('reveal_in_file_manager', { path });
}

export async function openInDefaultApp(path: string): Promise<void> {
  return invoke('open_in_default_app', { path });
}

export async function fileExists(path: string): Promise<boolean> {
  return invoke<boolean>('file_exists', { path });
}

export async function isDirectory(path: string): Promise<boolean> {
  return invoke<boolean>('is_directory', { path });
}

export async function getFileInfo(path: string): Promise<FileInfo> {
  return invoke<FileInfo>('get_file_info', { path });
}

// ============================================
// File Watcher
// ============================================

export async function watchPath(path: string): Promise<void> {
  return invoke('watch_path', { path });
}

export async function unwatchPath(path: string): Promise<void> {
  return invoke('unwatch_path', { path });
}

export async function onFileChange(
  callback: (event: FileChangeEvent) => void
): Promise<() => void> {
  return listen<FileChangeEvent>('file:change', callback);
}
